lib/puppet/provider/foreman_subnet/rest.rb

```diff
@@ -0,0 +1,263 @@
+  end
+
+  def domains=(value)
+    domain_ids = value.collect { |name| domain_id(name) }
+    subnets.update({'id' => id, 'subnet' => {'domain_ids' => domain_ids}})
+  end
+
+  def dhcp_proxy
+    subnet ? smartproxy_name(subnet['dhcp_id']) : nil
+  end
+
+  def dhcp_proxy=(value)
+    subnets.update({'id' => id, 'subnet' => {'dhcp_id' => smartproxy_id(value)}})
+  end
+
+  def tftp_proxy
+    subnet ? smartproxy_name(subnet['tftp_id']) : nil
+  end
+
+  def tftp_proxy=(value)
+    subnets.update({'id' => id, 'subnet' => {'tftp_id' => smartproxy_id(value)}})
+  end
+
+  def dns_proxy
+    subnet ? smartproxy_name(subnet['dns_id']) : nil
+  end
+
+  def dns_proxy=(value)
+    subnets.update({'id' => id, 'subnet' => {'dns_id' => smartproxy_id(value)}})
+  end
+
+  ## Helper functions
+  ##
+
+  private
+
+  def api_domains
+    @api_domains ||= ForemanApi::Resources::Domain.new({
+      :base_url => resource[:base_url],
+      :oauth    => {
+        :consumer_key    => resource[:consumer_key],
+        :consumer_secret => resource[:consumer_secret]
+      }
+    }, {
+      :headers => {
+        :foreman_user => resource[:effective_user]
+      }
+    })
+  end
+
+  def domain_name(id)
+    spi = api_domains.index[0]
+    if spi.is_a?(Hash) && spi.has_key?('results')
+      # foreman_api 0.1.18+
+      result = spi['results'].find { |s| s['id'] == id }
+    else
+      # foreman_api < 0.1.18
+      result = spi.find { |s| s['domain']['id'] == id }['domain']
+    end
+    result['name']
+  end
+
+  def domain_id(name)
+    spi = api_domains.index[0]
+    if spi.is_a?(Hash) && spi.has_key?('results')
+      # foreman_api 0.1.18+
+      result = spi['results'].find { |s| s['name'] == name }
+    else
+      # foreman_api < 0.1.18
+      result = spi.find { |s| s['domain']['name'] == name }['domain']
+    end
+    result['id']
+  end
+
+  def api_smartproxies
+    @api_smartproxies ||= ForemanApi::Resources::SmartProxy.new({
+      :base_url => resource[:base_url],
+      :oauth    => {
+        :consumer_key    => resource[:consumer_key],
+        :consumer_secret => resource[:consumer_secret]
+      }
+    }, {
+      :headers => {
+        :foreman_user => resource[:effective_user]
+      }
+    })
+  end
+
+  def smartproxy_id(name)
+    spi = api_smartproxies.index[0]
+    if spi.is_a?(Hash) && spi.has_key?('results')
+      # foreman_api 0.1.18+
+      result = spi['results'].find { |s| s['name'] == name }
+    else
+      # foreman_api < 0.1.18
+      result = spi.find { |s| s['smartProxy']['name'] == name }['smartProxy']
+    end
+    result['id']
+  end
+
+  def smartproxy_name(id)
+    spi = api_smartproxies.index[0]
+    if spi.is_a?(Hash) && spi.has_key?('results')
+      # foreman_api 0.1.18+
+      result = spi['results'].find { |s| s['id'] == id }
+    else
+      # foreman_api < 0.1.18
+      result = spi.find { |s| s['smartProxy']['id'] == id }['smartProxy']
```

Comment:
capitalisation here is odd, I wouldn't expect it, but this might go away with apipie-bindings :)